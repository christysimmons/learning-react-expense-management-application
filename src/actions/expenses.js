import uuid from 'uuid';
import db from '../firebase/firebase'

//No database action workflow

//component calls action generator
//action generator returns ---(expense) object---
//component dispatches object to store
//redux store updates with new object information

//Now with database action workflow

//component calls action generator
//action generator returns ---function---
//component dispatches (?) function ---middleware---
//function runs, updating store, adding to db, other actions as warranted


//Add Expense
export const addExpense = (expense) => ({
    type: 'ADD_EXPENSE',
    expense
}); 

//Add Expense database call

export const startAddExpense = (expenseData = {}) => {
    return (dispatch) => {
        const { 
            description = '', 
            note='', 
            amount = 0,
            createdAt = 0
        } = expenseData;
        const expense = { description, note, amount, createdAt }; 
        
        return db.ref('expenses').push(expense).then((ref) => {
            dispatch(addExpense({
                id: ref.key,
                ...expense
            }));
        })
    };
};

//Edit Expense

export const editExpense = (id, updates) => ({
    type: 'EDIT_EXPENSE', 
    id, 
    updates
});

//set_expenses from database to redux store

export const setExpenses = (expenses) => ({
    type: 'SET_EXPENSES', 
    expenses
});

export const startSetExpenses = () => {
    return (dispatch) => {
        return db.ref('expenses').once('value').then((snapshot) => {
            const expenses = [];
            snapshot.forEach((childSnapshot) => {
                expenses.push({
                    id: childSnapshot.key,
                    ...childSnapshot.val()
                });
            });
            dispatch(setExpenses(expenses));
        });
    };
};


//Delete Expense

export const removeExpense = ({ id }) => ({
    type: 'REMOVE_EXPENSE',
    id
});

export const startRemoveExpense = ({ id }) => {
    return (dispatch) => {
        return db.ref(`expenses/${id}`).remove().then(() => {
            dispatch(removeExpense({ id }));
        });
    }
};