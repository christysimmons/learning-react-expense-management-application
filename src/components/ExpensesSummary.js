import React from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import numeral from 'numeral';
import selectExpenses from '../selectors/expenses';
import getExpensesTotal from "../selectors/expenses-total";

/*
---First shot---

const ExpensesSummary = (props) => (
            <div>
            <p> 
                Viewing&nbsp;
                {props.expenses.length} 
                {(props.expenses.length === 1 ) ? " expense " : " expenses " }
                totaling&nbsp;
                {numeral(getExpensesTotal(props.expenses) / 100 ).format('$0,0.00')}.
            </p>
            </div>
);

const mapStateToProps = (state) => {
    return {
        expenses: selectExpenses(state.expenses, state.filters)
    };
};*/

//Second shot

export const ExpensesSummary = ({ expensesTotal, expenseCount }) => {
    const formatExpenseTotal = numeral(expensesTotal / 100 ).format('$0,0.00')
    return (
        <div className="page-header">
            <div className="content-container">  
                <h2 className="page-header__title">
                Viewing <span>{expenseCount}</span> {(expenseCount === 1 ) ? "expense" : "expenses" } totalling <span>{formatExpenseTotal}</span>.
                </h2>
                <div className="page-header__actions">
                <Link to="/create" className="button">Add a new expense </Link> 
                </div>
            </div>
        </div>
    )
};

const mapStateToProps = (state) => {
    const visibleExpenses = selectExpenses(state.expenses, state.filters);
    return {
        expenseCount: visibleExpenses.length,
        expensesTotal: getExpensesTotal(visibleExpenses)
    }

}

export default connect(mapStateToProps)(ExpensesSummary);