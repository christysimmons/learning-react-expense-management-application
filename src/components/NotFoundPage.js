import React from 'react';
import { Link } from 'react-router-dom';

const NotFoundPage = () => (
    <div>
        <p>404 not found. <Link to="/"> Go home instead?</Link></p>
    </div>
)

export default NotFoundPage;