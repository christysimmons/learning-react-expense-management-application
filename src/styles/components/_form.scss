.form {
    display: flex;
    flex-direction: column;
    > * {
        margin-bottom: $sm-size;
    }
    > button {
        width: 8.6rem;
    }
}

.form__error {
    margin: 0 0 $med-size 0;
    font-style: italic;
}