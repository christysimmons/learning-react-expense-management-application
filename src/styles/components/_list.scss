.list-header {
    background-color: $off-white;
    border: 1px solid darken($off-white, 7%);
    color: $gray;
    display: flex;
    justify-content: space-between;
    padding: $sm-size $med-size;
}

.list-body {
    margin-bottom: $med-size;
    @media (min-width: $desktop-breakpoint) {
        margin-bottom: $lg-size;
    }
}

.list-item {
    border: 1px solid darken($off-white, 7%);
    border-top: none;
    color: $darkgray;
    display: flex;
    flex-direction: column;
    padding: $sm-size;
    text-decoration: none;
    transition: background .3s ease;
    &:hover {
        background: $off-white;
    }
    @media (min-width: $desktop-breakpoint) {
        align-items: center;
        flex-direction: row;
        justify-content: space-between;
        padding: $med-size;
    }
}

.list-item--message {
    align-items: center;
    justify-content: center;
    color: $gray;
    padding: $med-size;
    &:hover {
        background: none;
    }
}

.list-item__title {
    margin: 0;
    word-break: break-all;
}

.list-item__subtitle {
    color: $gray;
    font-size: $font-size-small;
}

.list-item__amount {
    margin: $sm-size 0 0 0;
    @media (min-width: $desktop-breakpoint) {
        margin: 0;
        padding-left: $sm-size;
    }
}