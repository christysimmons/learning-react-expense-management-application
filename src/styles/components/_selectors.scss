.button {
    background-color: $blue;
    border: none;
    border-radius: 2px;
    box-shadow: none;
    -moz-box-shadow: none;
    -webkit-box-shadow: none;
    color: white;
    display: inline-block;
    font-size: $font-size-large;
    font-weight: 300;
    line-height: 1;
    outline: 0;
    padding: $sm-size;
    text-decoration: none;
}

.button--link {
    background-color: rgba(0,0,0,0);
}

.button--secondary {
    background-color: #888888;
}