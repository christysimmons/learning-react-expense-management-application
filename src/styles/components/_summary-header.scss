.page-header {
    background-color: $off-white;
    margin-bottom: $lg-size;
    padding: $lg-size 0;
}

.page-header__actions {
    margin-top: $med-size;
}

.page-header__title {
    font-weight: 300;
    margin: 0;
    span { 
        font-weight: bold;
    }
}