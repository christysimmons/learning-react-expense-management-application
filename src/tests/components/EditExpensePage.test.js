import React from 'react';
import { shallow } from 'enzyme';
import { EditExpensePage } from '../../components/EditExpensePage';
import expenses from '../fixtures/expenses';

let editExpense, removeExpense, startRemoveExpense, history, wrapper;

beforeEach(() => {
    editExpense = jest.fn();
    startRemoveExpense = jest.fn();
    history = { push: jest.fn() };
    wrapper = shallow(<EditExpensePage 
        editExpense={editExpense} 
        startRemoveExpense={startRemoveExpense} 
        history={history}
        expense={expenses[3]}
        />
    );
});

test('should render EditExpensePage', () => {
    expect(wrapper).toMatchSnapshot();
});

test('should handle onSubmit', () => {
    wrapper.find('ExpenseForm').prop('onSubmit')(expenses[3]);
    expect(history.push).toHaveBeenLastCalledWith('/');
    expect(editExpense).toHaveBeenLastCalledWith(expenses[3].id,expenses[3]);
});

test('should handle onClick', () => {
    wrapper.find('button').simulate('click');
    expect(history.push).toHaveBeenLastCalledWith('/');
    expect(startRemoveExpense).toHaveBeenLastCalledWith({
        id: expenses[3].id
    });
});