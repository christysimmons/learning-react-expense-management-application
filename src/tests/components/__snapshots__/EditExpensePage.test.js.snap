// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render EditExpensePage 1`] = `
<div>
  <div
    className="page-header"
  >
    <div
      className="content-container"
    >
      <h2
        className="page-header__title"
      >
         Edit an Expense 
      </h2>
    </div>
  </div>
  <div
    className="content-container"
  >
    <ExpenseForm
      expense={
        Object {
          "amount": 4400,
          "createdAt": 17276400000,
          "description": "Power bill",
          "id": "04",
          "note": "",
        }
      }
      onSubmit={[Function]}
    />
    <button
      className="button button--secondary"
      onClick={[Function]}
    >
       Remove 
    </button>
  </div>
</div>
`;
